examples/schroedinger-splitting/schroedinger-splitting.cpp

```diff
@@ -0,0 +1,202 @@
+
+*****************************<DESCRIPTION>**********************************/
+
+#include <ginkgo/ginkgo.hpp>
+
+#include <chrono>
+#include <fstream>
+#include <iostream>
+
+#include <opencv2/core.hpp>
+#include <opencv2/videoio.hpp>
+
+
+// This function implements a simple Ginkgo-themed clamped color mapping for
+// values in the range [0,5].
+void set_val(unsigned char *data, double value)
+{
+    // RGB values for the 6 colors used for values 0, 1, ..., 5
+    // We will interpolate linearly between these values.
+    double col_r[] = {255, 221, 129, 201, 249, 255};
+    double col_g[] = {255, 220, 130, 161, 158, 204};
+    double col_b[] = {255, 220, 133, 93, 24, 8};
+    value = std::max(0.0, value);
+    auto i = std::max(0, std::min(4, int(value)));
+    auto d = std::max(0.0, std::min(1.0, value - i));
+    // OpenCV uses BGR instead of RGB by default, revert indices
+    data[2] = static_cast<unsigned char>(col_r[i + 1] * d + col_r[i] * (1 - d));
+    data[1] = static_cast<unsigned char>(col_g[i + 1] * d + col_g[i] * (1 - d));
+    data[0] = static_cast<unsigned char>(col_b[i + 1] * d + col_b[i] * (1 - d));
+}
+
+
+// Initialize video output with given dimension and FPS (frames per seconds)
+std::pair<cv::VideoWriter, cv::Mat> build_output(int n, double fps)
+{
+    cv::Size videosize{n, n};
+    auto output =
+        std::make_pair(cv::VideoWriter{}, cv::Mat{videosize, CV_8UC3});
+    auto fourcc = cv::VideoWriter::fourcc('a', 'v', 'c', '1');
+    output.first.open("nls.mp4", fourcc, fps, videosize);
+    return output;
+}
+
+
+// Write the current frame to video output using the above color mapping
+void output_timestep(std::pair<cv::VideoWriter, cv::Mat> &output, int n,
+                     const std::complex<double> *data)
+{
+    for (int i = 0; i < n; i++) {
+        auto row = output.second.ptr(i);
+        for (int j = 0; j < n; j++) {
+            set_val(&row[3 * j], abs(data[i * n + j]));
+        }
+    }
+    output.first.write(output.second);
+}
+
+
+int main(int argc, char *argv[])
+{
+    using vec = gko::matrix::Dense<std::complex<double>>;
+    using real_vec = gko::matrix::Dense<double>;
+    using fft2 = gko::matrix::Fft2;
+
+    // Problem parameters:
+    // simulation length
+    auto t0 = 15.0;
+    // scaling factor for non-linearity
+    auto nonlinear_scale = 1.0;
+    // scaling factor for potential
+    auto potential_scale = 3.0;
+    // Simulation parameters:
+    // time scaling factor
+    auto time_scale = 0.25;
+    // number of grid points in each dimension
+    auto n = 256;
+    // number of simulation steps per second
+    auto steps_per_sec = 1000;
+    // number of video frames per second
+    auto fps = 25;
+    // number of grid points
+    auto n2 = n * n;
+    // phase difference between neighboring grid points
+    auto h = 2.0 * gko::pi<double>() / n;
+    auto h2 = h * h;
+    // time step size for the simulation
+    auto tau = 1.0 / steps_per_sec;
```

Comment:
Can you make all of them `const auto`, since you don't modify them (and the lambda `idx` would potentially output weird values)?